knimin/lib/vioscreen.py

```diff
@@ -0,0 +1,446 @@
+        for i in range(retries):
+            req = self._session.post(url, **kwargs)
+            if req.status_code != 200:  # HTTP status code, 200 is all good
+                data = req.json()
+
+                # if we did not get a HTTP status code 200, than guess that the
+                # API token is no longer valid so get a new one and retry
+                if 'Code' in data and data['Code'] == 1016:
+                    self._headers['token'] = self.get_token()
+                else:
+                    print(self._session.post, url, kwargs)
+                    raise ValueError("Unable to make this query work")
+            else:
+                return req.json()
+        raise ValueError("Unable to make this query work")
+
+    def tidyfy(self, username, payload):
+        """Restructures data so that 'survey_id' is associated with each row
+
+        Parameters
+        ----------
+        username: str
+            Survey ID that is being inserted each of the rows
+        payload: list of dict
+            The data that is getting associated with the survey_id
+
+        Return
+        ------
+        list of dict
+            The newly formatted data containing the username in each row
+        """
+        dat = []
+        for entry in payload:
+            entry['survey_id'] = username
+            dat.append(entry)
+        return dat
+
+    def get_session_data(self, session_id, endpoint):
+        """Pulls data from the vioscreen API based on 
+           a specific session ID and session type(ex. 'foodcomponents')
+
+        Parameters
+        ----------
+        session_id: str
+            Session ID that data is being requested for
+        endpoint: str
+            Name of the session type that data is being requested for
+
+        Return
+        ------
+        dict:
+            Food frequency questionnaire data
+        """
+        if not self._headers:
+            self._headers = {'Accept': 'application/json',
+                             'Authorization': 'Bearer %s' % self.get_token()}
+        return self.get('https://api.viocare.com/%s/sessions/%s/%s' %
+                         (self._key, session_id, endpoint),
+                         headers=self._headers)
+
+    def sync_vioscreen(self, user_ids=None):
+        """Pulls data from the vioscreen API and stores
+           the data into the AG database
+
+        Parameters
+        ----------
+        user_ids: set of str
+            Set of user_ids (identical to survey_ids) that
+            are needed to have their data pulled. Default None (syncs all)
+        """
+        if not self._headers:
+            self._headers = {'Accept': 'application/json',
+                             'Authorization': 'Bearer %s' % self.get_token()}
+        if not self._users:
+            self._users = self.get_users()
+
+        if user_ids:
+            if type(user_ids) is not set:
+                raise TypeError('user_ids should be type set')
```

Comment:
Since we control the signature of `get_users`, can't this constraint be enforced by the method such that it always returns a `set`?

Reply:
user_ids is referring to the input by the user, which is not explicitly set yet. If the input, user_ids is None, then user_ids will be set to all of the user_ids pullable from VioScreen like {'surveyid1', 'surveyid2', ...., 'surveyidn'}, where if you set user_ids to something like {'surveyid1'}, it'll just pull info for 'surveyid1'. self._users is different in that it contains users with a lot of extraneous data that isn't necessary for this script, so I use user_ids to filter the users that are needed for pulling.